### ISSUE TITLE
Rule Change: Support `language` and `dialects` to `meta`

### ISSUE THREAD

**Issue by lumirlumir** [eslint/markdown#339]
### What rule do you want to change?

Every rule

### What change do you want to make?

N/A

### How do you think the change should be implemented?

Other

### Example code

```markdown
meta: {
		type: "problem",

		docs: {
			description: "Disallow HTML tags",
		},

		messages: {
			disallowedElement: 'HTML element "{{name}}" is not allowed.',
		},

		schema: [
			{
				type: "object",
				properties: {
					allowed: {
						type: "array",
						items: {
							type: "string",
						},
						uniqueItems: true,
					},
				},
				additionalProperties: false,
			},
		],
	},
```

### What does the rule currently do for this code?

Currently, none of the rules have the `language` or `dialects` fields in their `meta`.

I noticed that all new rules added to ESLint for TypeScript support now include the `language` and `dialects` fields:

https://github.com/eslint/eslint/pull/19562/files#diff-d46d32957ba664c4b406c7587c5e047ab7f719ea98a597e0316837d61be4fc48R150-R151

Like we did for JavaScript, I’d like to suggest adding the `language` and `dialects` fields to `meta` as well.

### What will the rule do after it's changed?

I think we could add them like this:

```js
meta: {
    // ...
    language: 'markdown',
    dialects: ['commonmark', 'gfm'],
    // ...
}
```

The `language` field should be `'markdown'`, and the `dialects` can be `'commonmark'` and `'gfm'`.

### Participation

- [x] I am willing to submit a pull request to implement this change.

**Comment by nzakas** [comment]
We are holding off on doing this for other plugins as we are revisiting this design: 
https://github.com/eslint/eslint/issues/19462

**Comment by lumirlumir** [comment]
I'm copying and pasting the same concern about adding the `languages` field to the Markdown plugin from https://github.com/eslint/eslint/pull/20716#issuecomment-4239442753.

---

After further investigation of the language plugins, I found that adding a `languages` field could introduce breaking changes to those plugins.

For example, the Markdown plugin includes a rule called [`no-bare-urls`](https://github.com/eslint/markdown/blob/main/docs/rules/no-bare-urls.md). This rule works only in Markdown GFM mode, because the condition under which it reports an error is satisfied only when a GFM-compliant parser is used.

So, if we add a `languages` field, it should be `languages: ["markdown/gfm"]`, and `"markdown/commonmark"` should be omitted, since this error is never reported in CommonMark mode.

The problem is here. Since using this rule in CommonMark mode does not cause any harm, users may still enable it even when they are using CommonMark mode rather than GFM. Also, the current implementation of the rule visits nodes that can be generated in CommonMark mode, so the node listener functions can still run.

In conclusion, if a user is using the `no-bare-urls` rule in CommonMark mode, introducing `languages: ["markdown/gfm"]` would require them to change their configuration file.

So, if we want to add this field, I think it needs to be considered for the next major release, including in language plugins.

**Comment by lumirlumir** [comment]
I’m marking this PR as accepted, since JSON v2 has been released with this feature included, and there’s also an ongoing PR on the CSS side.

- https://github.com/eslint/json/pull/238
